File: 450/arrays/01_reverse.py
# approach 1 : iterative
# Time complexity: O(n)
# Space complexity: O(1)
def reverse(arr, start, end):
    while start < end:
        arr[start], arr[end] = arr[end], arr[start]
        start = start + 1
        end = end - 1
    return arr

# approach 2 : recursive
# Time complexity: O(n)
# Space complexity: O(n)
def reverse2(arr, start, end):
    if start > end:
        return
    reverse2(arr, start+1, end - 1)
    return arr


arr = [1, 2, 3, 4, 5]
print(reverse(arr, 0, len(arr) - 1))
print(reverse2(arr, 0, len(arr)-1))


# time comeplexities from shortest to longest
# O(1) < O(log n) < O(n) < O(n log n) < O(n^2) < O(2^n) < O(n!)

# space complexities from shortest to longest
# O(1) < O(log n) < O(n) < O(n log n) < O(n^2) < O(2^n) < O(n!)

File: javascript/450/arrays/01-reverse.js
// approach 1 : create a new array and loop through the array from the end and push the elements to the new array
// Time complexity : O(n)
// Space complexity : O(n)

function reverse(arr) {
  let newArr = [];
  for (let i = arr.length - 1; i >= 0; i--) {
    newArr.push(arr[i]);
  }
  return newArr;
}

// approach 2 : swap the elements from the start and end
// Time complexity : O(n)
// Space complexity : O(1)
function reverse2(arr) {
  let start = 0;
  let end = arr.length - 1;
  while (start < end) {
    let temp = arr[start];
    arr[start] = arr[end];
    arr[end] = temp;
    start++;
    end--;
  }

  return arr;
}


// approach 3 : recursive approach
// Time complexity : O(n)
// Space complexity : O(n), because it creates a new stack frame for each recursive call
function reverse3(arr, start, end){
    if(start >= end){
        return arr;
    }
    let temp = arr[start];
    arr[start] = arr[end];
    arr[end] = temp;
    return reverse3(arr, start+1, end-1);

}


console.log(reverse([1, 2, 3, 4, 5]));
console.log(reverse2([1, 2, 3, 4, 5]));
console.log(reverse3([1, 2, 3, 4, 5], 0, 4));
